plugins/externalkeyboard.koplugin/find-keyboard.lua

```diff
@@ -0,0 +1,102 @@
+local bit = require("bit")
+local ffi = require("ffi")
+local lfs = require("libs/libkoreader-lfs")
+
+-- Constants from the linux kernel input-event-codes.h
+local KEY_UP = 103
+local BTN_DPAD_UP = 0x220
+
+local FindKeyboard = {}
+
+local function count_set_bits(n)
+    -- Brian Kernighan's algorithm
+    local count = 0
+    while n ~= 0 do
+        count = count + 1
+        n = bit.band(n, n - 1)
+    end
+    return count
+end
+
+local function capabilities_str_to_long_bitmap_array(str)
+    -- The format for capabilities is at include/linux/mod_devicetable.h.
+    -- They are long's split by spaces. See linux/drivers/input/input.c::input_print_bitmap.
+    local long_bitmap_arr = {}
+    for c in str:gmatch "([0-9a-fA-F]+)" do
+        local long_bitmap = tonumber(c, 16)
+        table.insert(long_bitmap_arr, 1, long_bitmap)
+    end
+    return long_bitmap_arr
+end
+
+local function count_set_bits_in_array(arr)
+    local count = 0
+    for __, number in ipairs(arr) do
+        local count_in_number = count_set_bits(number)
+        count = count + count_in_number
+    end
+
+    return count
+end
+
+local function is_capabilities_bit_set(long_bitmap_arr, bit_offset)
+    local long_bitsize = ffi.sizeof("long") * 8
+    local arr_index = math.floor(bit_offset / long_bitsize)
+    local long_mask = bit.lshift(1, bit_offset % long_bitsize)
+    local long_bitmap = long_bitmap_arr[arr_index + 1] -- Array index starts from 1 in Lua
+
+    if long_bitmap then
+        return bit.band(long_bitmap, long_mask) ~= 0
+    else
+        return false
+    end
+end
+
+local function read_key_capabilities(sys_event_path)
+    local key_path = sys_event_path .. "/device/capabilities/key"
+    local file = io.open(key_path, "r")
+    if not file then
+        -- This should not happen - the kernel creates key capabilities file for all devices.
+        return nil
+    end
+    local keys_bitmap_str = file:read("l")
+    file:close()
+
+    return capabilities_str_to_long_bitmap_array(keys_bitmap_str)
+end
+
+local function analyze_key_capabilities(long_bitmap_arr)
+    -- The heuristic is that a keyboard has at least as many keys as there are alphabet letters and some more.
+    local keyboard_min_number_keys = 64
```

Comment:
This may be too few. When developing this feature I assumed the use case of a user carrying an ereader and a compact keyboard. A 40% keyboard has ~45-50 keys. I agree that 28 is too few, though.

Can you post capabilities and modalias of the device that was detected as a keyboard?

Reply:
It stood at 98 (or maybe that's my local keyboard, can't remember ^^).

While, yeah, it was a bit too low, It didn't really help anyway, as the extra evdev device created with my test keyboard *also* exposes a crapload of keys (slightly smaller that the main one, true, but hard to discriminate heuristically nonetheless).

So, basically, that 64 came from counting the amount of keys in the central "letters+numbers+spacebar" block of a standard keyboard (which tallies at ~62 w/o function keys).

Reply:
```
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(130:2.07:12%:03:58:72%:#)── cat /sys/class/input/event4/device/modalias                                                                                                                                                                                                                                                                             ──(Sat, Oct 29)─┘
input:b0003v1532p0118e0111-e0,1,4,11,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,8C,8E,96,98,9E,9F,A1,A3,A4,A5,A6,AD,B0,B1,B2,B3,B4,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,F0,ram4,l0,1,2,sfw
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.07:12%:03:58:72%:#)── cat /sys/class/input/event4/device/capabilities/key                                                                                                                                                                                                                                                                         ──(Sat, Oct 29)─┘
10000 7 ff9f207a c14057ff febeffdf ffefffff ffffffff fffffffe
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.07:12%:03:58:72%:#)── cat /sys/class/input/event5/device/modalias                                                                                                                                                                                                                                                                                 ──(Sat, Oct 29)─┘
input:b0003v1532p0118e0111-e0,1,2,3,4,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,8B,8C,8E,8F,90,96,98,9B,9C,9E,9F,A1,A3,A4,A5,A6,A7,A8,A9,AB,AC,AD,AE,B0,B1,B2,B3,B4,B5,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,CE,CF,D0,D1,D2,D4,D8,D9,DB,E0,E1,E4,EA,EB,F0,F1,F4,100,161,162,166,16A,16E,172,174,176,178,179,17A,17B,17C,17D,17F,180,182,183,185,188,189,18C,18D,18E,18F,190,191,192,193,195,197,198,199,19A,19C,1A0,1A1,1A2,1A3,1A4,1A5,1A6,1A7,1A8,1A9,1AA,1AB,1AC,1AD,1AE,1AF,1B0,1B1,1B7,1BA,240,241,242,243,244,245,246,250,251,r6,a20,28,29,2A,2B,2C,2D,2E,2F,30,31,32,33,34,35,36,37,38,39,3A,3B,3C,3D,3E,m4,lsfw
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.06:12%:03:58:72%:#)── cat /sys/class/input/event5/device/capabilities/key                                                                                                                                                                                                                                                                         ──(Sat, Oct 29)─┘
3007f 0 0 0 0 483ffff 17aff32d bf544446 0 0 1 130c13 b17c007 ffbf7bfa d941dfff febeffdf ffefffff ffffffff fffffffe
```

`event4` is the one that actually reports keypresses.

Reply:
Hmm, is there any difference between `/sys/class/input/event[45]/device/properties`?

Reply:
Nope:

```
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.13:13%:07:59:71%:#)── cat /sys/class/input/event5/device/properties                                                                                                                                                                                                                                                                               ──(Sat, Oct 29)─┘
0
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.12:13%:07:59:71%:#)── cat /sys/class/input/event4/device/properties                                                                                                                                                                                                                                                                               ──(Sat, Oct 29)─┘
0
```

Reply:
event4 is the only one with LEDs, though...

```
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.13:13%:08:01:71%:#)── ls -lash /sys/class/input/event5/device/                                                                                                                                                                                                                                                                                    ──(Sat, Oct 29)─┘
total 0      
     0 drwxr-xr-x    6 admin    root           0 Oct 29 07:59 .
     0 drwxr-xr-x    3 admin    root           0 Oct 29 07:59 ..
     0 drwxr-xr-x    2 admin    root           0 Oct 29 07:59 capabilities
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 07:59 device -> ../../../0003:1532:0118.0035
     0 drwxr-xr-x    3 admin    root           0 Oct 29 07:59 event5
     0 drwxr-xr-x    2 admin    root           0 Oct 29 07:59 id
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 modalias
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 name
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 phys
     0 drwxr-xr-x    2 admin    root           0 Oct 29 07:59 power
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 properties
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 07:59 subsystem -> ../../../../../../../../../../class/input
     0 -rw-r--r--    1 admin    root        4.0K Oct 29 07:59 uevent
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 uniq
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.12:13%:08:01:71%:#)── ls -lash /sys/class/input/event4/device/                                                                                                                                                                                                                                                                                    ──(Sat, Oct 29)─┘
total 0      
     0 drwxr-xr-x    9 admin    root           0 Oct 29 07:59 .
     0 drwxr-xr-x    3 admin    root           0 Oct 29 07:59 ..
     0 drwxr-xr-x    2 admin    root           0 Oct 29 07:59 capabilities
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 08:00 device -> ../../../0003:1532:0118.0034
     0 drwxr-xr-x    3 admin    root           0 Oct 29 07:59 event4
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:00 id
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:00 input55::capslock
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:00 input55::numlock
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:00 input55::scrolllock
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:00 modalias
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:00 name
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:00 phys
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:00 power
     0 -r--r--r--    1 admin    root        4.0K Oct 29 07:59 properties
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 07:59 subsystem -> ../../../../../../../../../../class/input
     0 -rw-r--r--    1 admin    root        4.0K Oct 29 07:59 uevent
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:00 uniq
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.11:13%:08:01:71%:#)── cat /sys/class/input/event4/device/uevent                                                                                                                                                                                                                                                                                   ──(Sat, Oct 29)─┘
PRODUCT=3/1532/118/111
NAME="Razer Razer DeathStalker"
PHYS="usb-sunxi-ohci-1/input0"
UNIQ=""
PROP=0
EV=120013
KEY=10000 7 ff9f207a c14057ff febeffdf ffefffff ffffffff fffffffe
MSC=10
LED=7
MODALIAS=input:b0003v1532p0118e0111-e0,1,4,11,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,8C,8E,96,98,9E,9F,A1,A3,A4,A5,A6,AD,B0,B1,B2,B3,B4,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,F0,ram4,l0,1,2,sfw
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.09:13%:08:01:71%:#)── cat /sys/class/input/event5/device/uevent                                                                                                                                                                                                                                                                                   ──(Sat, Oct 29)─┘
PRODUCT=3/1532/118/111
NAME="Razer Razer DeathStalker"
PHYS="usb-sunxi-ohci-1/input1"
UNIQ=""
PROP=0
EV=10001f
KEY=3007f 0 0 0 0 483ffff 17aff32d bf544446 0 0 1 130c13 b17c007 ffbf7bfa d941dfff febeffdf ffefffff ffffffff fffffffe
REL=40
ABS=ffffff01 0
MSC=10
MODALIAS=input:b0003v1532p0118e0111-e0,1,2,3,4,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,8B,8C,8E,8F,90,96,98,9B,9C,9E,9F,A1,A3,A4,A5,A6,A7,A8,A9,AB,AC,AD,AE,B0,B1,B2,B3,B4,B5,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,CE,CF,D0,D1,D2,D4,D8,D9,DB,E0,E1,E4,EA,EB,F0,F1,F4,100,161,162,166,16A,16E,172,174,176,178,179,17A,17B,17C,17D,17F,180,182,183,185,188,189,18C,18D,18E,18F,190,191,192,193,195,197,198,199,19A,19C,1A0,1A1,1A2,1A3,1A4,1A5,1A6,1A7,1A8,1A9,1AA,1AB,1AC,1AD,1AE,1AF,1B0,1B1,1B7,1BA,240,241,242,243,244,245,246,250,251,r6,a20,28,29,2A,2B,2C,2D,2E,2F,30,31,32,33,34,35,36,37,38,39,3A,3B,3C,3D,3E,m4,lsfw
```

Reply:
That holds true on my desktop, too, FWIW:

```
┌─(niluje@illyria:pts/18)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.12:30%:08:02:%)── ls -lash /sys/class/input/event{9..12}/device/                                                                                                                                                                                                                                                                                  ──(Sat, Oct 29)─┘
/sys/class/input/event10/device/:
total 0
0 drwxr-xr-x 6 root root    0 Oct 27 00:12 .
0 drwxr-xr-x 6 root root    0 Oct 27 00:05 ..
0 drwxr-xr-x 2 root root    0 Oct 28 04:33 capabilities
0 lrwxrwxrwx 1 root root    0 Oct 29 08:02 device -> ../../../0003:1B1C:1B55.0001
0 drwxr-xr-x 3 root root    0 Oct 27 00:12 event10
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 id
0 -rw-r--r-- 1 root root 4.0K Oct 29 08:02 inhibited
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 modalias
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 name
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 phys
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 power
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 properties
0 lrwxrwxrwx 1 root root    0 Oct 27 00:12 subsystem -> ../../../../../../../../../class/input
0 -rw-r--r-- 1 root root 4.0K Oct 27 00:12 uevent
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 uniq

/sys/class/input/event11/device/:
total 0
0 drwxr-xr-x 6 root root    0 Oct 27 00:12 .
0 drwxr-xr-x 6 root root    0 Oct 27 00:05 ..
0 drwxr-xr-x 2 root root    0 Oct 28 04:33 capabilities
0 lrwxrwxrwx 1 root root    0 Oct 29 08:02 device -> ../../../0003:1B1C:1B55.0001
0 drwxr-xr-x 3 root root    0 Oct 27 00:12 event11
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 id
0 -rw-r--r-- 1 root root 4.0K Oct 29 08:02 inhibited
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 modalias
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 name
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 phys
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 power
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 properties
0 lrwxrwxrwx 1 root root    0 Oct 27 00:12 subsystem -> ../../../../../../../../../class/input
0 -rw-r--r-- 1 root root 4.0K Oct 27 00:12 uevent
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 uniq

/sys/class/input/event12/device/:
total 0
0 drwxr-xr-x 8 root root    0 Oct 27 00:05 .
0 drwxr-xr-x 6 root root    0 Oct 27 00:05 ..
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 capabilities
0 lrwxrwxrwx 1 root root    0 Oct 29 08:02 device -> ../../../0003:1B1C:1B55.0001
0 drwxr-xr-x 3 root root    0 Oct 27 00:05 event12
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 id
0 -rw-r--r-- 1 root root 4.0K Oct 29 08:02 inhibited
0 drwxr-xr-x 3 root root    0 Oct 27 00:12 js0
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 modalias
0 drwxr-xr-x 3 root root    0 Oct 27 00:12 mouse0
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 name
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 phys
0 drwxr-xr-x 2 root root    0 Oct 29 08:02 power
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 properties
0 lrwxrwxrwx 1 root root    0 Oct 27 08:11 subsystem -> ../../../../../../../../../class/input
0 -rw-r--r-- 1 root root 4.0K Oct 27 00:12 uevent
0 -r--r--r-- 1 root root 4.0K Oct 29 08:02 uniq

/sys/class/input/event9/device/:
total 0
0 drwxr-xr-x 9 root root    0 Oct 27 00:05 .
0 drwxr-xr-x 6 root root    0 Oct 27 00:05 ..
0 drwxr-xr-x 2 root root    0 Oct 28 23:14 capabilities
0 lrwxrwxrwx 1 root root    0 Oct 28 23:14 device -> ../../../0003:1B1C:1B55.0001
0 drwxr-xr-x 3 root root    0 Oct 27 00:05 event9
0 drwxr-xr-x 2 root root    0 Oct 28 23:14 id
0 -rw-r--r-- 1 root root 4.0K Oct 28 23:14 inhibited
0 drwxr-xr-x 3 root root    0 Oct 28 23:14 input9::capslock
0 drwxr-xr-x 3 root root    0 Oct 28 23:14 input9::numlock
0 drwxr-xr-x 3 root root    0 Oct 28 23:14 input9::scrolllock
0 -r--r--r-- 1 root root 4.0K Oct 28 23:14 modalias
0 -r--r--r-- 1 root root 4.0K Oct 28 23:14 name
0 -r--r--r-- 1 root root 4.0K Oct 28 23:14 phys
0 drwxr-xr-x 2 root root    0 Oct 28 23:14 power
0 -r--r--r-- 1 root root 4.0K Oct 28 23:14 properties
0 lrwxrwxrwx 1 root root    0 Oct 27 08:11 subsystem -> ../../../../../../../../../class/input
0 -rw-r--r-- 1 root root 4.0K Oct 27 00:12 uevent
0 -r--r--r-- 1 root root 4.0K Oct 28 23:14 uniq
┌─(niluje@illyria:pts/18)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.34:30%:08:03:%)── cat /sys/class/input/event{9..12}/device/uevent                                                                                                                                                                                                                                                                                 ──(Sat, Oct 29)─┘
PRODUCT=3/1b1c/1b55/111
NAME="Corsair CORSAIR K70 RGB MK.2 LOW PROFILE Mechanical Gaming Keyboard"
PHYS="usb-0000:00:14.0-3/input0"
UNIQ="1001701DAF1E1C455A6FC297F5001BC5"
PROP=0
EV=12001f
KEY=3f000307ff 0 0 483ffff17aff32d bfd4444600000000 1 130ff38b17d007 ffe77bfad9415fff ffbeffdfffefffff fffffffffffffffe
REL=1040
ABS=100000000
MSC=10
LED=7
MODALIAS=input:b0003v1B1Cp1B55e0111-e0,1,2,3,4,11,14,k71,72,73,74,75,77,78,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,8B,8C,8E,90,96,98,9B,9C,9E,9F,A1,A3,A4,A5,A6,A7,A8,A9,AB,AC,AD,AE,B0,B1,B2,B5,B6,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,CC,CE,CF,D0,D1,D2,D4,D8,D9,DB,DF,E0,E1,E4,E5,E6,E7,E8,E9,EA,EB,F0,F1,F4,100,161,162,166,16A,16E,172,174,176,177,178,179,17A,17B,17C,17D,17F,180,182,183,185,188,189,18C,18D,18E,18F,190,191,192,193,195,197,198,199,19A,19C,1A0,1A1,1A2,1A3,1A4,1A5,1A6,1A7,1A8,1A9,1AA,1AB,1AC,1AD,1AE,1AF,1B0,1B1,1B7,1BA,240,241,242,243,244,245,246,247,248,249,24A,250,251,260,261,262,263,264,265,r6,C,a20,m4,l0,1,2,sfw
PRODUCT=3/1b1c/1b55/111
NAME="Corsair CORSAIR K70 RGB MK.2 LOW PROFILE Mechanical Gaming Keyboard"
PHYS="usb-0000:00:14.0-3/input0"
UNIQ="1001701DAF1E1C455A6FC297F5001BC5"
PROP=0
EV=9
ABS=10000000000
MODALIAS=input:b0003v1B1Cp1B55e0111-e0,3,kra28,mlsfw
PRODUCT=3/1b1c/1b55/111
NAME="Corsair CORSAIR K70 RGB MK.2 LOW PROFILE Mechanical Gaming Keyboard"
PHYS="usb-0000:00:14.0-3/input0"
UNIQ="1001701DAF1E1C455A6FC297F5001BC5"
PROP=0
EV=9
ABS=10000000000
MODALIAS=input:b0003v1B1Cp1B55e0111-e0,3,kra28,mlsfw
PRODUCT=3/1b1c/1b55/111
NAME="Corsair CORSAIR K70 RGB MK.2 LOW PROFILE Mechanical Gaming Keyboard Mouse"
PHYS="usb-0000:00:14.0-3/input0"
UNIQ="1001701DAF1E1C455A6FC297F5001BC5"
PROP=0
EV=17
KEY=ffffffff0000 0 0 0 0
REL=903
MSC=10
MODALIAS=input:b0003v1B1Cp1B55e0111-e0,1,2,4,k110,111,112,113,114,115,116,117,118,119,11A,11B,11C,11D,11E,11F,120,121,122,123,124,125,126,127,128,129,12A,12B,12C,12D,12E,12F,r0,1,8,B,am4,lsfw
```

Reply:
Ditto with the aforementioned old Logitech:

```
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.01:14%:08:07:70%:#)── cat /sys/class/input/event{4..5}/device/uevent                                                                                                                                                                                                                                                                              ──(Sat, Oct 29)─┘
PRODUCT=3/46d/c30e/110
NAME="Logitech HID compliant keyboard"
PHYS="usb-sunxi-ohci-1/input0"
UNIQ=""
PROP=0
EV=120013
KEY=10000 7 ff800000 7ff febeffdf f3cfffff ffffffff fffffffe
MSC=10
LED=7
MODALIAS=input:b0003v046DpC30Ee0110-e0,1,4,11,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,F0,ram4,l0,1,2,sfw
PRODUCT=3/46d/c30e/110
NAME="Logitech HID compliant keyboard"
PHYS="usb-sunxi-ohci-1/input1"
UNIQ=""
PROP=0
EV=13
KEY=fff ffffffff ffffffff 0 2000000 387a d800d001 1e0000 0 0 0
MSC=10
MODALIAS=input:b0003v046DpC30Ee0110-e0,1,4,k71,72,73,74,80,8C,8E,8F,9B,9C,9E,9F,A1,A3,A4,A5,A6,AB,AC,AD,D9,100,101,102,103,104,105,106,107,108,109,10A,10B,10C,10D,10E,10F,110,111,112,113,114,115,116,117,118,119,11A,11B,11C,11D,11E,11F,120,121,122,123,124,125,126,127,128,129,12A,12B,12C,12D,12E,12F,130,131,132,133,134,135,136,137,138,139,13A,13B,13C,13D,13E,13F,140,141,142,143,144,145,146,147,148,149,14A,14B,ram4,lsfw
┌─(ROOT@europa:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(2.01:14%:08:07:70%:#)── ls -lash /sys/class/input/event{4..5}/device/                                                                                                                                                                                                                                                                               ──(Sat, Oct 29)─┘
/sys/class/input/event4/device/:
total 0      
     0 drwxr-xr-x    9 admin    root           0 Oct 29 08:07 .
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:07 ..
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:07 capabilities
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 08:08 device -> ../../../0003:046D:C30E.0039
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:07 event4
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:08 id
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:08 input60::capslock
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:08 input60::numlock
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:08 input60::scrolllock
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 modalias
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 name
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 phys
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:08 power
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 properties
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 08:07 subsystem -> ../../../../../../../../../../class/input
     0 -rw-r--r--    1 admin    root        4.0K Oct 29 08:07 uevent
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 uniq

/sys/class/input/event5/device/:
total 0      
     0 drwxr-xr-x    6 admin    root           0 Oct 29 08:07 .
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:07 ..
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:07 capabilities
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 08:08 device -> ../../../0003:046D:C30E.003A
     0 drwxr-xr-x    3 admin    root           0 Oct 29 08:07 event5
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:08 id
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 modalias
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 name
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 phys
     0 drwxr-xr-x    2 admin    root           0 Oct 29 08:08 power
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 properties
     0 lrwxrwxrwx    1 admin    root           0 Oct 29 08:07 subsystem -> ../../../../../../../../../../class/input
     0 -rw-r--r--    1 admin    root        4.0K Oct 29 08:07 uevent
     0 -r--r--r--    1 admin    root        4.0K Oct 29 08:08 uniq
```

Reply:
Ev seems to differ significantly. I'll see if it can be used for classification.